### ISSUE TITLE
Inconsistency regarding whether a variable annotated with a TypeVar bound to an abstract class can be instantiated

### ISSUE THREAD

**Issue by davidatbu** [microsoft/pyright#2018]
### Discussed in https://github.com/microsoft/pyright/discussions/2017
**Describe the bug**
 It would be much easier to describe succinctly after presenting the reproduction below.

**To Reproduce**
Consider the following:

```python
from abc import ABC, abstractmethod
from typing import Type, TypeVar

_ASub = TypeVar("_ASub", bound="A")
class A(ABC):
    @abstractmethod
    def some_method(self) -> str: ...

    def some_factory_method(self: _ASub) -> _ASub:
        return type(self)() # Pyright complains that `A` is abstract


class ActualASub(A):
    def some_method(self) -> str:
        return "some string"


ActualASub().some_factory_method()
```
This runs fine, but pyright complains.

If this is by design, then the following should also raise an error, but it doesn't.

```py
# Use the code snippet above, but replace `some_factory_method` with the following classmethod.
    @classmethod
    def some_factory_method(cls: Type[_ASub]) -> _ASub:
        return cls()
```
**Expected behavior**
Either none or both of the snippets must pass type checking. But currently, the first one doesn't and the second one does.

**VS Code extension or command-line**
This behavior was observed with the command line invocation of pyright, version 1.1.140.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.152, which I just published. It will also be included in the next release of pylance.
